[1.3] Implement user namespaces

alban opened the pull request:
This is the containerd/cri implementation for the [Kubernetes Node-Level User Namespaces Design Proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/node-usernamespace-remapping.md).

The patches are based on the `release/1.3` branch. It is tested on Kubernetes 1.17 with patches adapted from PR 64005 (https://github.com/kinvolk/kubernetes/pull/3).

The main changes are:
- Import the new CRI API from Kubernetes PR 64005.
- Implement `GetRuntimeConfigInfo` returning the hard coded uid mapping with 100000
- Use the `WithRemappedSnapshot` snapshotter
- Fix sysfs mount with correct ownership of netns (see commitmsg for details)
- Additional mount restrictions on /dev/shm (nosuid, noexec, nodev)
- Chown /dev/shm appropriately
- Fix etc-hosts mounts with supplementary groups (see commitmsg for details)
- Use custom options `WithoutNamespace`, `WithUserNamespace`, `WithLinuxNamespace` at the right places

At the OCI level (config.json), we have the following changes:
- sandbox container with **new** "user" namespace and **UidMappings**
- normal containers with "user" namespace from a **path** and **UidMappings**

Demo:

```
$ kubectl apply -f userns-tests/pod-simple.yaml 
pod/pod-simple created
$ kubectl apply -f userns-tests/pod-userns.yaml
pod/pod-userns created
$ kubectl exec -ti pod-simple -- /bin/sh -c 'cat /proc/self/uid_map'
         0          0 4294967295
$ kubectl exec -ti pod-userns -- /bin/sh -c 'cat /proc/self/uid_map'
         0     100000      65535
```

<details>

* userns-tests/pod-simple.yaml

```
apiVersion: v1
kind: Pod
metadata:
  name: pod-simple
  namespace: default
  annotations:
    alpha.kinvolk.io/userns: "disabled"
spec:
  containers:
  - name: pod-simple
    image: busybox
    command: ["sh"]
    args: ["-c", "sleep infinity"]
```

* userns-tests/pod-userns.yaml:

```
apiVersion: v1
kind: Pod
metadata:
  name: pod-userns
  namespace: default
  annotations:
    alpha.kinvolk.io/userns: "enabled"
spec:
  containers:
  - name: pod-userns
    image: busybox
    command: ["sh"]
    args: ["-c", "sleep infinity"]
```

</details>

TODO:
- [x] Make the uid mapping configurable instead of hard coding 100000.
- [ ] Try on the master branch (presumably upstream wants an implementation on master first)
- [ ] Tests.
- [ ] Implement `NamespaceMode=NODE_WIDE_REMAPPED` correctly. At the moment, `NamespaceMode=POD` and `NamespaceMode=NODE` are implemented correctly.
- [ ] Check the status of volumes.

/cc @mauriciovasquezbernal

alban committed:
Add new CRI API, GetRuntimeConfigInfo() and a new LinuxNamepsaceOption, User

Initial work by vikaschoudhary16 <vichoudh@redhat.com> on Kubernetes PR
64005.

alban committed:
Implement GetRuntimeConfigInfo()

With this patch, GetRuntimeConfigInfo() returns a config without user
namespace: the only uid/gid mapping is the one from the root user
namespace.

alban committed:
Fix typo on RunPodSandbox

alban committed:
Statically define a userns config

alban committed:
Implement sandbox with user namespaces

alban committed:
User namespaces: use the correct snapshotter

alban committed:
User namespaces: fix ownership of netns

Before this patch, containerd created a netns, configured it with CNI,
and then creates the sandbox container by giving the netns path previously
setup. This means that the netns was owned by the host userns. Mounting
sysfs in the container is restricted in this setup.

This patch sets up the netns in the other way around instead: it creates
the sandbox container, letting runc create a new netns. Then, it picks
the new netns from /proc/$pid/ns/net, binds mount it in the usual CNI
path and then gives it to CNI to configure. This means that the netns is
owned by the userns of the sandbox container. In this way, mounting
sysfs is possible.

For more information about namespace ownership, see
- man ioctl_ns
- man user_namespaces, section "Interaction of user namespaces and other types of namespaces"
- Linux commit 7dc5dbc879bd ("sysfs: Restrict mounting sysfs")
  https://github.com/torvalds/linux/commit/7dc5dbc879bd0779924b5132a48b731a0bc04a1e#diff-4839664cd0c8eab716e064323c7cd71fR1164
- net_current_may_mount() used for mounting sysfs:
  ns_capable(net->user_ns, CAP_SYS_ADMIN);
  https://github.com/torvalds/linux/blob/v5.7/net/core/net-sysfs.c#L1679

alban committed:
User namespace: restrict mount options for sandbox /dev/shm (tmpfs)

The sandbox container (aka "pause" container) has a tmpfs mount on
/dev/shm. Bind mount it with nosuid, noexec, nodev because the mount
would not be allowed in user namespaces otherwise.

alban committed:
User namespaces: use the correct snapshotter

rata commented on pkg/server/sandbox_run.go:
```diff
@@ -499,6 +552,10 @@ func (c *criService) setupSandboxFiles(id string, config *runtime.PodSandboxConf
 			return errors.Wrap(err, "failed to mount sandbox shm")
 		}
+		uid := int(UsernsMapping.HostID)
+		if err := os.Chown(sandboxDevShm, uid, uid); err != nil {
```
Missing gid, I guess?

alban replied:
Right. It works because of how I hardcoded the values in pkg/server/runtime_config_info.go but I should change that to avoid forgetting to add the gid there.

rata commented on pkg/server/sandbox_run.go:
```diff
@@ -190,13 +151,31 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
+	case runtime.NamespaceMode_NODE:
+		snapshotterOption = customopts.WithNewSnapshot(id, containerdImage)
+	case runtime.NamespaceMode_NODE_WIDE_REMAPPED:
+		fallthrough
```
Not sure we want this option or not, though. For now, I'd not implement until talking with upstream. That option, IIUC, made sense when no containr runtime supported userns remapping, and now they do. So... maybe it is not important anymore?

alban replied:
Ok... Meanwhile, I will keep the CRI gRPC changes, and return an error if `NamespaceMode_NODE_WIDE_REMAPPED` is used, so that it does not silently implement `POD` when the kubelet asks for `NODE_WIDE_REMAPPED`.

mauriciovasquezbernal commented on pkg/netns/netns.go:
```diff
@@ -156,14 +164,29 @@ func getCurrentThreadNetNSPath() string {
 // NewNetNS creates a network namespace.
 func NewNetNS() (*NetNS, error) {
-	path, err := newNS()
+	path, err := newNS(0)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to setup netns")
+	}
+	return &NetNS{path: path}, nil
```
```suggestion
	return NetNSFromPID(0)
```

?

alban committed:
User namespaces: add SupplementaryGroups=0 in systemd units

runc needs to bind mount files from /var/lib/kubelet/pods/... (such as
etc-hosts) into the container. When using user namespaces, the bind
mount didn't work anymore when containerd is started from a systemd unit.

This patch fixes that by adding SupplementaryGroups=0

runc needs to have permission on the directory to stat() the source of
the bind mount. Without user namespaces, this is not a problem since
runc is running as root, so it has 'rwx' permissions over the directory:

drwxr-x---. 8 root   root   4096 May 28 18:05 /var/lib/kubelet

Moreover, runc has CAP_DAC_OVERRIDE at this point because the mount
phase happens before giving up the additional permissions.

However, when using user namespaces, the runc process is belonging to a
different user than root (depending on the mapping). /var/lib/kubelet is
seen as belonging to the special unmapped user (65534, nobody). runc
does not have 'rwx' permissions anymore but the empty '---' permission
for 'other'.

CAP_DAC_OVERRIDE is also no effective because the kernel performs the
capability check with capable_wrt_inode_uidgid(inode, CAP_DAC_OVERRIDE).
This checks that the owner of the /var/lib/kubelet is mapped in the
current user namespace, which is not the case.

Despite that, bind mounting /var/lib/kubelet/pods/...etc-hosts was
working when containerd was started manually with 'sudo' but not working
when started from a systemd unit. The difference is how supplementary
groups are handled between sudo and systemd units: systemd does not set
supplementary groups by default.

$ sudo grep -E 'Groups:|Uid:|Gid:' /proc/self/status
Uid:	0	0	0	0
Gid:	0	0	0	0
Groups:	0

$ sudo systemd-run -t grep -E 'Groups:|Uid:|Gid:' /proc/self/status
Running as unit: run-u296886.service
Press ^] three times within 1s to disconnect TTY.
Uid:	0	0	0	0
Gid:	0	0	0	0
Groups:

When runc has the supplementary group 0 configured, it is retained
during the bind-mount phase, even though it is an unmapped group (runc
temporarily sees 'Groups: 65534' in its own /proc/self/status), so runc
effectively has the 'r-x' permissions over /var/lib/kubelet. This makes
the bind mount of etc-hosts work.

After the mount phase, runc will set the credential correctly (following
OCI's config.json specification), so the container will not retain this
unmapped supplementary group.

I rely on the systemd unit file being configured correctly with
SupplementaryGroups=0 and I don't attempt to set it up automatically
with syscall.Setgroups() because "at the kernel level, user IDs and
group IDs are a per-thread attribute" (man setgroups) and the way Golang
uses threads make it difficult to predict which thread is going to be
used to execute runc. glibc's setgroup() is a wrapper that changes the
credentials for all threads but Golang does not use the glibc
implementation.

alban committed:
User namespaces: make ID mapping configurable

Example of possible configuration:
```
[plugins]
  [plugins."io.containerd.grpc.v1.cri"]
    [plugins."io.containerd.grpc.v1.cri".node_wide_uid_mapping]
      container_id = 0
      host_id = 300000
      size = 65536
    [plugins."io.containerd.grpc.v1.cri".node_wide_gid_mapping]
      container_id = 0
      host_id = 300000
      size = 65536
```

alban committed:
User namespaces: implement NamespaceMode NODE and POD

Return an error if NODE_WIDE_REMAPPED is requested

alban committed:
Tests for user namespaces

alban committed:
Fixup move userns options together

mauriciovasquezbernal commented on pkg/config/config.go:
```diff
@@ -271,6 +284,10 @@ func DefaultConfig() PluginConfig {
 		},
 		MaxConcurrentDownloads: 3,
 		DisableProcMount:       false,
+		//NodeWideUIDMapping:     LinuxIDMapping{0, 0, 4294967295},
+		//NodeWideGIDMapping:     LinuxIDMapping{0, 0, 4294967295},
+		NodeWideUIDMapping: LinuxIDMapping{0, 100000, 65536},
+		NodeWideGIDMapping: LinuxIDMapping{0, 100000, 65536},
```
```suggestion
		NodeWideUIDMapping:     LinuxIDMapping{0, 0, 4294967295},
		NodeWideGIDMapping:     LinuxIDMapping{0, 0, 4294967295},
```

?

mauriciovasquezbernal commented on pkg/config/config_test.go:
```diff
@@ -320,6 +320,47 @@ func TestValidateConfig(t *testing.T) {
+				NodeWideGIDMapping: LinuxIDMapping{1, 100000, 65536},
+			},
+			expectedErr: "missing root id in container",
+		},
```
I implemented a test for different uid/gid mapping in https://github.com/kinvolk/containerd-cri/commit/76ade408a665afacb92b25c28407f22d35aa7ff2, feel free to cherry-pick if you think that's good enough.

mauriciovasquezbernal commented on pkg/server/container_create.go:
```diff
@@ -175,6 +175,24 @@ func (c *criService) CreateContainer(ctx context.Context, r *runtime.CreateConta
+		return nil, errors.New("unsupported user namespace mode: CONTAINER")
+	case runtime.NamespaceMode_NODE:
+		snapshotterOption = customopts.WithNewSnapshot(id, containerdImage)
+	case runtime.NamespaceMode_NODE_WIDE_REMAPPED:
+		fallthrough
```
```suggestion
```

mauriciovasquezbernal commented on pkg/server/container_create.go:
```diff
@@ -438,6 +455,25 @@ func (c *criService) generateContainerSpec(id string, sandboxID string, sandboxP
+	switch securityContext.GetNamespaceOptions().GetUser() {
+	case runtime.NamespaceMode_NODE:
+		specOpts = append(specOpts, customopts.WithoutNamespace(runtimespec.UserNamespace))
+	case runtime.NamespaceMode_NODE_WIDE_REMAPPED:
+		fallthrough
```
```suggestion
```

mauriciovasquezbernal commented on pkg/server/instrumented_service.go:
```diff
@@ -441,6 +441,23 @@ func (in *instrumentedService) Status(ctx context.Context, r *runtime.StatusRequ
+		if err != nil {
+			log.G(ctx).WithError(err).Error("GetRuntimeConfigInfo failed")
+		} else {
+			log.G(ctx).Tracef("GetRuntimeConfigInfo returns status %+v", res.GetRuntimeConfig())
```
```suggestion
			log.G(ctx).Tracef("GetRuntimeConfigInfo returns %+v", res.GetRuntimeConfig())
```

mauriciovasquezbernal commented on pkg/server/sandbox_run.go:
```diff
@@ -190,13 +151,29 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
+		return nil, errors.New("unsupported user namespace mode: CONTAINER")
+	case runtime.NamespaceMode_NODE:
+		snapshotterOption = customopts.WithNewSnapshot(id, containerdImage)
+	case runtime.NamespaceMode_NODE_WIDE_REMAPPED:
+		return nil, errors.New("unsupported user namespace mode: NODE_WIDE_REMAPPED")
```
```suggestion
```

mauriciovasquezbernal commented on pkg/server/sandbox_run.go:
```diff
@@ -387,6 +410,26 @@ func (c *criService) generateSandboxContainerSpec(id string, config *runtime.Pod
+	switch nsOptions.GetUser() {
+	case runtime.NamespaceMode_NODE:
+		// nothing to do: defaultUnixNamespaces() already comes without user namespaces
+	case runtime.NamespaceMode_NODE_WIDE_REMAPPED:
+		return nil, errors.New("unsupported user namespace mode: NODE_WIDE_REMAPPED")
```
```suggestion
```

mauriciovasquezbernal reviewed (commented):
Some minor comments. I think we could remove NODE_WIDE_REMAPPED everywhere on this PR as we removed that on the kubelet one.
